### DISCUSSION TITLE
Large performance discrepancy between RLS performance in API vs SQL

### DISCUSSION THREAD

**Discussion by pranavbadami** [supabase/supabase#6127] [Questions] [answered]
Hey folks, I'm experiencing a strange discrepancy in RLS performance on an instance when I'm running an RPC via the API vs raw SQL.

I'm running an RPC called `active_users` with the same params on both API and in SQL. 

On API, I'm seeing average time to return of **around 30 seconds**. This is done via POST request to `rest/v1/rpc/active_users`. 

On raw SQL, I'm seeing closer to **7 seconds**. This is done via `SELECT * FROM active_users(...)` after I mock an authenticated session in psql with:
`SET ROLE authenticated;`
`SELECT set_config('request.jwt.claims','{"app_metadata": {"workspace": "...."}}', true);`

The total size of the data returned via the API is only ~80kb so the issue here is time to first byte, not download time. 

Does anyone have any pointers on what could be going on? Thanks in advance! 🙌

**Comment by GaryAustin1** [comment]
Do they return the same results?
Unless you set jwt.claims for sub and role those will be null if your RLS uses auth.uid() or auth.role().  Note auth.role() uses jwt claim, not postgres role.

**Comment by steve-chavez** [comment]
> On raw SQL, I'm seeing closer to 7 seconds. This is done via SELECT * FROM active_users(...) after I mock an authenticated session in psql with:
SET ROLE authenticated;
SELECT set_config('request.jwt.claims','{"app_metadata": {"workspace": "...."}}', true);
SELECT * FROM active_users(...)

We compared the simple query above(`SELECT * from active_users(..)`) with the PostgREST generated query(gotten from the logs) a and the latter one was actually faster for about 1 second.

Then I also compared the same inputs through an API call and the API overhead(latency) was ~1 second compared to plain SQL through `pgsql`.

> On API, I'm seeing average time to return of around 30 seconds. This is done via POST request to rest/v1/rpc/active_users.

The above was never confirmed but the suspicion was that a different JWT was being used as an input(yet not shared by @pranavbadami).

**Comment by steve-chavez** [comment, marked as answer]
Cross posting the solution to RLS perf with custom JWT claims: https://github.com/orgs/supabase/discussions/9311#discussioncomment-5690939
